commit 7ee2cb7f32b299c2b06a31fde155457203e4b7dd
Author: Frank Filz <ffilzlnx@us.ibm.com>
Date:   Mon May 18 17:41:40 2009 -0400

    nfs: Fix NFS v4 client handling of MAY_EXEC in nfs_permission.
    
    The problem is that permission checking is skipped if atomic open is
    possible, but when exec opens a file, it just opens it O_READONLY which
    means EXEC permission will not be checked at that time.
    
    This problem is observed by the following sequence (executed as root):
    
      mount -t nfs4 server:/ /mnt4
      echo "ls" >/mnt4/foo
      chmod 744 /mnt4/foo
      su guest -c "mnt4/foo"
    
    Signed-off-by: Frank Filz <ffilzlnx@us.ibm.com>
    Signed-off-by: Trond Myklebust <Trond.Myklebust@netapp.com>
    Cc: stable@kernel.org
    Tested-by: Eugene Teo <eugeneteo@kernel.sg>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Backported to Debian's 2.6.18 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.18.orig/fs/nfs/dir.c linux-source-2.6.18/fs/nfs/dir.c
--- linux-source-2.6.18.orig/fs/nfs/dir.c	2009-05-20 14:12:22.000000000 -0600
+++ linux-source-2.6.18/fs/nfs/dir.c	2009-08-10 23:43:40.000000000 -0600
@@ -1722,7 +1722,8 @@ int nfs_permission(struct inode *inode, 
 			/* NFSv4 has atomic_open... */
 			if (nfs_server_capable(inode, NFS_CAP_ATOMIC_OPEN)
 					&& nd != NULL
-					&& (nd->flags & LOOKUP_OPEN))
+					&& (nd->flags & LOOKUP_OPEN)
+					&& !(mask & MAY_EXEC))
 				goto out;
 			break;
 		case S_IFDIR:
